### PR TITLE
fix Issue 20855 - stack overflow when compiling large file

### PR DESCRIPTION
**Pull request by WalterBright**

### CODE REVIEW

**Review by Geod24 — DISMISSED** [review]
Since the code is so trivial, I think it'd be better to just write a script generating it.
Also this doesn't really fix the issue, just adds a hard-coded limit which works on your system.

**Comment by WalterBright** [comment]
Yes, switching the test case to be generative is better; done.

The fundamental problem here is the backend is recursive in nature. It's difficult to figure out just where to put the checks without engaging in whack-a-mole. The check I put in is conservative, it needs to go double that before it stack overflows. You're doing something wrong anyway with 10,000 statements in a function.

I also put the check in the backend rather than the frontend, as other compilers have their own means of dealing with the problem.

**Comment by ibuclaw** [comment]
Isn't the issue that your doing something quadratic when traversing trees in the dmd backend?  Why can't it be fixed to not be quadratic?  (i.e: isn't this just [issue 6401](https://issues.dlang.org/show_bug.cgi?id=6401#c6))

**Comment by WalterBright** [comment]
Quadratic doesn't cause stack overflow. Recursive depth does. Trying different statement counts in the attached test clearly indicate a stack overflow in optelem(), and optelem() is recursive.

There are many quadratic algorithms in the optimizer, which is why it is slow.

**Comment by WalterBright** [comment]
Don't approve yet. I'm going to try out another idea.

**Comment by WalterBright** [comment]
I completely changed the approach. Instead of connecting ExpStatements with OPcomma, it creates a new block for each ExpStatement. It resolves the stack overflow by changing the data structure from recursive to iterative. But it does not resolve the quadratic behavior in the optimizer. I cut the iterations in the test case from 100,000 to around 8,000 so it does not bog down the test suite.

There are already separate bugzilla issues for the quadratic behavior, so I won't concern this PR with them.

**Comment by WalterBright** [comment]
The test failures have something to do with -preview=fieldwise. Will investigate more tomorrow.

**Comment by WalterBright** [comment]
This PR uncovered an unrelated optimizer bug. I disabled the problem code, and will fix the optimizer in a followup PR.

**Comment by WalterBright** [comment]
Ubuntu test fails with:
```
CI agent stopped responding!
```

**Comment by WalterBright** [comment]
Windows_VisualD_LDC x86-mscoff failed

```
"D:\a\1\s\generated\Windows\RelWithAsserts\Win32\dmd.exe" -conf= -g -m32mscoff -O -w -preview=dip1000 -I../druntime\import -unittest -version=StdUnittest -version=CoreUnittest -c  -ofunittest2a.obj std\array.d  std\functional.d  std\path.d  std\outbuffer.d  std\utf.d

core.exception.AssertError@..\dmd\backend\cgcod.d(1045): Assertion failure
```
I have no idea why this should fail only when built with LDC?

**Comment by WalterBright** [comment]
buildkite AuburnSounds fails with:
```
core.exception.AssertError@source/inteli/xmmintrin.d(1154): unittest failure
```
Who is in charge of AuburnSounds?

**Comment by thewilsonator** [comment]
Guillaume Piolat @p0nce

**Comment by WalterBright** [comment]
@p0nce please come help with https://github.com/dlang/dmd/pull/12409

**Comment by p0nce** [comment]
The test that fail is:
```d
/// Extract a 16-bit unsigned integer from `a`, selected with `imm8`. Zero-extended.
int _mm_extract_pi16 (__m64 a, int imm8)
{
    short4 sa = cast(short4)a;
    return cast(ushort)(sa.array[imm8]);
}
unittest
{
    __m64 A = _mm_setr_pi16(-1, 6, 0, 4);
    assert(_mm_extract_pi16(A, 0) == 65535);
    assert(_mm_extract_pi16(A, 1) == 6);
    assert(_mm_extract_pi16(A, 2) == 0);
    assert(_mm_extract_pi16(A, 3) == 4);
}
```
(not reduced, vectors here are actually custom structs)

and it pass in DMD for a long time. It's not the fault of the test if it fails...

**Comment by WalterBright** [comment]
Thank you, @p0nce 

Can you turn it into a compilable example, along with what switches you use?

**Comment by p0nce** [comment]
Sure, wait about 30min

**Comment by p0nce** [comment]
I think this should reproduce your bug, build with `dmd -unittest -O -inline main.d`

Note: `intel-intrinsics` expose a lot of DMD backend bugs because:  1. sometimes I try to use D_SIMD  2. unittest have optimizations on   3. all sorts of basic stuff and operations. I spent a lot of efforts workarounding and reporting DMD backend bugs while developing this library, so it's very worthwhile to have it in the testsuite imho to avoid putting in more bugs.

**Comment by WalterBright** [comment]
Thank you. I can repro it.

I've been looking through the backend bugs,

https://issues.dlang.org/buglist.cgi?bug_status=NEW&bug_status=ASSIGNED&bug_status=REOPENED&keywords=backend&keywords_type=allwords&list_id=235749&query_format=advanced

and fixing them. Hence this PR. If you've filed ones that aren't on this list, please tag them with the `backend` keyword.

**Comment by p0nce** [comment]
Absolutely I do appreciate all the work done to make the backend rock solid.

File: src/dmd/backend/cgcod.d [hunk 1]
```diff
@@ -1034,10 +1039,27 @@ else
                 anyiasm ||
                 Alloca.size
                )
+            {
                 needframe = 1;
+                if (!(mfuncreg & mBP))
+                {
+                    fprintf(stderr, "%s %d %d %d %d %d x%x\n",
```
Comment (ibuclaw) [inline src/dmd/backend/cgcod.d:1046]: Debug printf needs to be commented out before this goes in.

**Comment by WalterBright** [comment]
@Geod24  This is ready to go now, but pull https://github.com/dlang/dmd/pull/12444 first. (This incorporates the changes from 12444 so it'll pass the test suite.)

The test suite failures are the usual heisenbugs which are plaguing the test suites.

**Comment by WalterBright** [comment]
@Geod24 @RazvanN7 this is ready to go.

File: src/dmd/backend/blockopt.d [hunk 2]
```diff
@@ -660,13 +660,19 @@ void blockopt(int iter)
             blexit();
             if (iter >= 2)
                 brmin();                // minimize branching
+            version (MARS)
+                // Switched to one block per Statement, do not undo it
+                enum merge = false;
+            else
+                enum merge = true;
+
```
Comment (Geod24) [inline src/dmd/backend/blockopt.d:668]: This looks like it's going to come back to haunt you down the road 🤷

**Review by Geod24 — APPROVED** [review]

**Comment by WalterBright** [comment]
Yay!

**Comment by MoonlightSentinel** [comment]
Maybe related to these changes?
https://issues.dlang.org/show_bug.cgi?id=22098

**Comment by CyberShadow** [comment]
This pull request introduced a regression:
https://issues.dlang.org/show_bug.cgi?id=22175

**Comment by ibuclaw** [comment]
This PR introduced a regression https://issues.dlang.org/show_bug.cgi?id=23084

**Pull request merged**